components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOFacade.java

```diff
@@ -0,0 +1,313 @@
+            throws ActionMgtException {
+
+        if (updatingActionDTO.getEndpoint() == null || updatingActionDTO.getEndpoint().getAuthentication() == null) {
+            return;
+        }
+
+        Authentication updatingAuthentication = updatingActionDTO.getEndpoint().getAuthentication();
+        Authentication existingAuthentication = existingActionDTO.getEndpoint().getAuthentication();
+
+        try {
+            if (updatingAuthentication.getType() != existingAuthentication.getType()) {
+                actionSecretProcessor.deleteAssociatedSecrets(existingAuthentication, existingActionDTO.getId());
+            }
+            List<AuthProperty> encryptedProperties = actionSecretProcessor.encryptAssociatedSecrets(
+                    updatingAuthentication, updatingActionDTO.getId());
+            updatingActionDTO.setAuthenticationProperties(encryptedProperties);
+        } catch (SecretManagementException e) {
+            throw new ActionMgtServerException("Error while updating Action Endpoint Authentication Secrets.", e);
+        }
+    }
+
+    private void deleteAuthenticationSecrets(ActionDTO deletingActionDTO) throws ActionMgtServerException {
+
+        try {
+            actionSecretProcessor.deleteAssociatedSecrets(deletingActionDTO.getEndpoint().getAuthentication(),
+                    deletingActionDTO.getId());
+        } catch (SecretManagementException e) {
+            throw new ActionMgtServerException("Error while deleting Action Endpoint Authentication Secrets.", e);
+        }
+    }
+
+    private void addProperties(ActionDTO actionDTO, Integer tenantId) throws ActionPropertyResolverException {
+
+        Map<String, String> properties = null;
+        ActionPropertyResolver actionPropertyResolver =
+                ActionPropertyResolverFactory.getActionPropertyResolver(actionDTO.getType());
+
+        if (actionPropertyResolver != null) {
+            properties = actionPropertyResolver.addProperties(actionDTO,
+                    IdentityTenantUtil.getTenantDomain(tenantId));
+        }
+        if (properties != null) {
+            actionDTO.setProperties(properties.entrySet().stream()
+                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)));
+        }
+    }
+
+    private void getPropertiesOfActionDTOs(String actionType, List<ActionDTO> actionDTOS, Integer tenantId)
+            throws ActionPropertyResolverException {
+
+        ActionPropertyResolver actionPropertyResolver =
+                ActionPropertyResolverFactory.getActionPropertyResolver(
+                        org.wso2.carbon.identity.action.management.model.Action.ActionTypes.valueOf(actionType));
+        if (actionPropertyResolver == null) {
+            return;
+        }
+
+        for (ActionDTO actionDTO : actionDTOS) {
+            actionDTO.setProperties(actionPropertyResolver.getProperties(actionDTO,
+                    IdentityTenantUtil.getTenantDomain(tenantId)));
+        }
```

Comment:
We should have a way to let the downstream component to populate the ActionDTOs in a bulk manner. If there are any I/O operations this pattern will be heavy

Reply:
Fixed with Fixed with https://github.com/wso2/carbon-identity-framework/pull/6119/commits/dab9dd75643167cc04aa5d35aec2f044d59a58d4